lib/straight/blockchain_adapters_dispatcher.rb

```diff
@@ -1,60 +1,67 @@
 module Straight
 
   # Employs many adapters and runs requests simultaneously.
-  # How many adapters are employed is determined by STEP.
+  # How many adapters are employed is determined by var passed to #initializer (default 2).
   # If all adapters fail it raises an exception.
   class BlockchainAdaptersDispatcher
 
-    STEP = 2
-    attr_reader :list_position, :adapters, :result, :step
+    class AdaptersTimeoutError < TimeoutError; end
+    class AdaptersError < StandardError; end
 
-    def initialize(adapters, &block)
-      @list_position = 0
-      @result        = nil
-      @step          = 0
-      @adapters = adapters
+    TIMEOUT = 60
+    attr_reader :list_position, :adapters, :result, :defer_result, :step, :tasks_parallel_limit
+
+    def initialize(adapters, tasks_parallel_limit: 2, &block)
+      @list_position        = 0
+      @defer_result         = Concurrent::IVar.new
+      @tasks_parallel_limit = tasks_parallel_limit
+      @step                 = 0
+      @adapters             = adapters
       run_requests(block) if block_given?
     end
 
     def get_adapters
-      @step = [adapters.size - @list_position, STEP].min
-      adapters = @adapters[@list_position..@step-1]
+      @step = [adapters.size - @list_position, @tasks_parallel_limit].min
+      adapters = @adapters[@list_position...@list_position+@step]
       @list_position += @step
       adapters
     end
 
     def run_requests(block)
-      threads = []
-      fault_counter = 0
-      get_adapters.each do |adapter| 
-        t = Thread.new do
-          block.call(adapter)
-        end
-        t.abort_on_exception = true
-        threads << t
-      end
-      wait_values(threads)
-    rescue => e
-      raise e if fault_counter == @step && @list_position == @adapters.size
-      if fault_counter == @step 
-        fault_counter = 0
-        retry
-      end
-      fault_counter += 1
+      execute_in_parallel(block, get_adapters)
+      Timeout::timeout(TIMEOUT, AdaptersTimeoutError) {
+        @result = @defer_result.wait.value
+      }
     end
+    
+  private
 
-    def wait_values(threads)
-      threads.each do |t| 
-        Thread.new do
-          if val = t.value
-            @result = val
-            threads.map(&:kill)
-          end
+    def execute_in_parallel(block, adapters)
+      attempts = Concurrent::MVar.new(0)
+      pool = Concurrent::ThreadPoolExecutor.new
```

Comment:
Maybe this should be `FixedThreadPool.new(@tasks_parallel_limit)`?
Then you can just put all adapters in a queue and do not mess with the `@list_position` and `@step`.